lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp

```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+  int B = opTensorShape[dim.batch];
+  int K = opTensorShape[dim.k];
+  int NonK = opTensorShape[dim.nonK];
+
+  auto shapePerCTATile =
+      expandMatrixShapeWithBatch(ArrayRef(getShapePerCTATile(dLayout)));
+  shapePerCTATile[dim.k] = K;
+  auto sizePerThread =
+      expandMatrixShapeWithBatch(ArrayRef(getSizePerThread(dLayout)));
+  sizePerThread[dim.k] = K;
+  auto threadsPerWarp =
+      expandMatrixShapeWithBatch(ArrayRef(dLayout.getThreadsPerWarp()));
+  auto warpsPerCTA =
+      expandMatrixShapeWithBatch(ArrayRef(dLayout.getWarpsPerCTA()));
+
+  auto warpSize = i32_val(triton::gpu::getWarpSize(dLayout));
+  auto laneId = urem(thread, warpSize);
+  auto warpId = udiv(thread, warpSize);
+  auto laneIds =
+      mlir::LLVM::delinearize(rewriter, loc, laneId, threadsPerWarp, opOrder);
+  auto warpIds =
+      mlir::LLVM::delinearize(rewriter, loc, warpId, warpsPerCTA, opOrder);
+  auto sizePerWarpB = sizePerThread[dim.batch] * threadsPerWarp[dim.batch];
+  auto sizePerWarpNonK = sizePerThread[dim.nonK] * threadsPerWarp[dim.nonK];
+
+  Value bTileOffset =
+      mul(laneIds[dim.batch], i32_val(sizePerThread[dim.batch]));
+  bTileOffset =
+      add(bTileOffset, mul(warpIds[dim.batch], i32_val(sizePerWarpB)));
+  Value nonKTileOffset =
+      mul(laneIds[dim.nonK], i32_val(sizePerThread[dim.nonK]));
+  nonKTileOffset =
+      add(nonKTileOffset, mul(warpIds[dim.nonK], i32_val(sizePerWarpNonK)));
+
+  auto elemTy = typeConverter->convertType(opTensorTy.getElementType());
+  Type ptrTy = smem.base.getType();
+
+  auto sharedOrder = expandMatrixOrderWithBatch(sharedLayout.getOrder());
+  // compute contiguity of fastest dimension in shared layout.
+  unsigned vectorSize = sizePerThread[sharedOrder[0]];
+  if (sharedLayout.getMaxPhase() > 1)
+    vectorSize = std::min(vectorSize, sharedLayout.getVec());
+  auto vecTy = vec_ty(elemTy, vectorSize);
+  // loop increments depend on fastest dim
+  unsigned dimStep[3] = {1, 1, 1};
+  dimStep[sharedOrder[0]] = vectorSize;
+
+  auto shapePerCTABTile = shapePerCTATile[dim.batch];
+  auto shapePerCTANonKTile = shapePerCTATile[dim.nonK];
+  auto sizeBPerThread = sizePerThread[dim.batch];
+  auto sizeNonKPerThread = sizePerThread[dim.nonK];
+  auto numBTiles = std::max(1u, B / shapePerCTABTile);
+  auto numNonKTiles = std::max(1u, NonK / shapePerCTANonKTile);
+
+  auto perThreadShape =
+      getElemsPerThreadInOp(opTensorShape, shapePerCTATile, sizePerThread);
+
+  SmallVector<Value> opValues(numBTiles * sizeBPerThread * K * numNonKTiles *
+                              sizeNonKPerThread);
+
+  bool swizzlePath = isSwizzled(sharedLayout);
+
+  // In swizzled memory case basePtr stores pointer to the beginning of shared
+  // memmory object.
+  //
+  // If memory is not swizzled, algorithm breaks element offset pointer into
+  // constant and non-constant part. Non-constant (depends on thread id) part is
+  // same for all elements, so it is computed only once. basePtr stores this
+  // non-constant part
```

Comment:
"Non-constant (depends on thread id) part is same for all elements,"
->
"Non-constant (depends on thread id) part is the offset of the first element of the thread, which is same for all elements of the thread."